src/gen_bpf.c

```diff
@@ -1222,111 +1229,257 @@ static struct bpf_blk *_gen_bpf_syscall(struct bpf_state *state,
 }
 
 /**
- * Generate the BPF instruction blocks for a given filter/architecture
- * @param state the BPF state
- * @param db the filter DB
- * @param db_secondary the secondary DB
+ * Calculate the number of levels in the binary tree
+ * @param syscall_cnt the number of syscalls in this seccomp filter
+ */
+static int _get_bintree_levels(unsigned int syscall_cnt,
+			       uint32_t optimize)
+{
+	unsigned int i = 2, max_level = SYSCALLS_PER_NODE * 2;
+
+	if (syscall_cnt < MIN_SYSCALLS_TO_USE_BINTREE ||
+	    optimize != 2)
+		/* Only use a binary tree if there are a lot of syscalls
+		 * and if the filter optimization level is set to 2 (aka
+		 * binary tree)
+		 */
+		return 0;
+
+	while(max_level < syscall_cnt) {
+		max_level <<= 1;
+		i++;
+	}
+
+	return i;
```

Comment:
Perhaps I'm missing some finer point, but can't you simply divide by two here(`return syscall_cnt / 2;`)?

Reply:
Hmmm... I'm not sure.  I remember this function was trickier to write than I expected, and I grudgingly settled on this as it seemed the best balance of readability and speed.  I'll look through it again.